pkg/controller/deployment/deployment_controller.go

```diff
@@ -148,3 +198,142 @@ func (d *DeploymentController) getDesiredRC(deployment *experimental.Deployment)
+	totalReplicaCount := 0
+	for _, rc := range replicationControllers {
+		totalReplicaCount += rc.Spec.Replicas
+	}
+	return totalReplicaCount
+}
+
+func (d *DeploymentController) scaleUp(allRCs []*api.ReplicationController, desiredRC *api.ReplicationController, deployment experimental.Deployment) (bool, error) {
+	maxSurge, isPercent, err := util.GetIntOrPercentValue(&deployment.Spec.Strategy.RollingUpdate.MaxSurge)
+	if err != nil {
+		return false, fmt.Errorf("Invalid value for MaxSurge: %v", err)
+	}
+	if isPercent {
+		maxSurge = util.GetValueFromPercent(maxSurge, deployment.Spec.Replicas)
+	}
+	// Find the total number of pods
+	allPods, err := d.getPodsForRCs(allRCs)
+	if err != nil {
+		return false, err
+	}
+	currentPodCount := len(allPods)
+	// Check if we can scale up.
+	maxTotalPods := deployment.Spec.Replicas + maxSurge
+	if currentPodCount >= maxTotalPods {
+		// Cannot scale up.
+		return false, nil
+	}
+	// Scale up.
+	scaleUpCount := maxTotalPods - currentPodCount
+	scaleUpCount = int(math.Min(float64(scaleUpCount), float64(deployment.Spec.Replicas-desiredRC.Spec.Replicas)))
+	_, err = d.scaleRC(desiredRC, scaleUpCount)
+	return true, err
+}
+
+func (d *DeploymentController) scaleDown(allRCs []*api.ReplicationController, targetedRCs []*api.ReplicationController, desiredRC *api.ReplicationController, deployment experimental.Deployment) (bool, error) {
+	maxUnavailable, isPercent, err := util.GetIntOrPercentValue(&deployment.Spec.Strategy.RollingUpdate.MaxUnavailable)
+	if err != nil {
+		return false, fmt.Errorf("Invalid value for MaxUnavailable: %v", err)
+	}
+	if isPercent {
+		maxUnavailable = util.GetValueFromPercent(maxUnavailable, deployment.Spec.Replicas)
+	}
+	// Check if we can scale down.
+	minAvailable := deployment.Spec.Replicas - maxUnavailable
+	// Find the number of ready pods.
+	// TODO: Use MinReadySeconds once https://github.com/kubernetes/kubernetes/pull/12894 is merged.
+	readyPodCount := 0
+	allPods, err := d.getPodsForRCs(allRCs)
+	for _, pod := range allPods {
+		if api.IsPodReady(&pod) {
+			readyPodCount++
+		}
+	}
+
+	if readyPodCount <= minAvailable {
+		// Cannot scale down.
+		return false, nil
+	}
+	totalScaleDownCount := readyPodCount - minAvailable
+	for _, targetRC := range targetedRCs {
+		if totalScaleDownCount == 0 {
+			// No further scaling required.
+			break
+		}
+		if targetRC.Spec.Replicas == 0 {
+			// cannot scale down this RC.
+			continue
+		}
+		// Scale down.
+		scaleDownCount := int(math.Min(float64(targetRC.Spec.Replicas), float64(totalScaleDownCount)))
+		_, err = d.scaleRC(targetRC, -scaleDownCount)
+		if err != nil {
+			return false, err
+		}
+		totalScaleDownCount -= scaleDownCount
+	}
+	return true, err
+}
+
+func (d *DeploymentController) updateDeploymentStatus(allRCs []*api.ReplicationController, desiredRC *api.ReplicationController, deployment experimental.Deployment) error {
```

Comment:
The comment on Replicas and UpdatedReplicas currently say this:

```
	// Total number of ready pods targeted by this deployment (this
	// includes both the old and new pods).
	Replicas int `json:"replicas,omitempty"`

	// Total number of new ready pods with the desired template spec.
	UpdatedReplicas int `json:"updatedReplicas,omitempty"`
```

But this code doesn't talk about ready pods, it just adds the RC replica counts regardless of whether that many pods actually exist. Either we should change this code to reflect ready pods, change the comments, or put a todo that we're going to reconcile towards the comments.